### ISSUE TITLE
NVPIPE_H264_FFMPEG decoder fails to decode valid frames returning NVPIPE_EOVERFLOW

### ISSUE THREAD

**Issue by renewagner** [NVIDIA/NvPipe#9]
Hi,

I'm trying to implement a remote visualization client/server based on `egl-example.cpp` with an `NVPIPE_H264_NV` encoder on the server and an `NVPIPE_H264_FFMPEG` decoder on clients with pre-Kepler GPU hardware. After applying the patch in #8 I can compile NvPipe with `USE_FFMPEG=ON` but the decoder returns `NVPIPE_EOVERFLOW`. My code (derived from `egl-example.cpp`) is as follows:

```
#include <cstring>
#include <string>
#include <iostream>
#include <fstream>

#include <nvpipe.h>

#include "mapped_vector.hpp"

void writePPM(const uint8_t *rgb, const uint32_t width, const uint32_t height, const std::string& path)
{
    // For verification...

    size_t numBytes = width * height * 3;

    std::ofstream outFile;
    outFile.open(path.c_str(), std::ios::binary);

    outFile << "P6" << "\n"
            << width << " " << height << "\n"
            << "255\n";

    outFile.write((char*) rgb, numBytes);
}


int main(int /*argc*/, char** /*argv*/)
{
    const uint32_t width = 1920;
    const uint32_t height = 1080;

    mapped_vector<uint8_t> f("frame.h264");

    std::cout << "input has " << f.size() << " bytes" << std::endl;

    /*
     * Client: Init decoder
     */
    nvpipe* decoder = nvpipe_create_decoder(NVPIPE_H264_FFMPEG);
    
    size_t clientBufferSize = width * height * 3;
    uint8_t *clientBuffer = new uint8_t[clientBufferSize];

    nvp_err_t decodeStatus = nvpipe_decode(decoder, &*f.begin(), f.size(), clientBuffer, width, height, NVPIPE_RGB);
    if (decodeStatus != NVPIPE_SUCCESS)
    {
        std::cerr << "Decode failed: " << std::string(nvpipe_strerror(decodeStatus)) << std::endl;
        return 1;
    }

    writePPM(clientBuffer, width, height, "frame.ppm");

    delete[] clientBuffer;
    nvpipe_destroy(decoder);

    return 0;
}
```

The `width` and `height` parameters are identical on the server. `mapped_vector` is just a wrapper around `mmap(2)` that allows access to a binary file using a `std::vector`-like (read-only) API.

The same code works fine if I change the decoder to `NVPIPE_H264_NV` (on a machine with a Pascal GPU). I can also view the file just fine with `ffplay -f h264 -framerate 30 -i frame.h264` (on the same machine as the code above).

Any suggestions?

Cheers,

Rene

**Comment by tfogal** [comment]
Hi René, thanks for the ping.
As mentioned in the readme, we do not support pre-Kepler hardware in NvPipe at this time.  There may just be a simple bug in how the decoder is created, or it may be more involved.  Another alternative would be to just use ffmpeg's API directly: ffmpeg has had support for NVIDIA hardware for quite some time.

Apologies!

**Comment by renewagner** [comment]
Hi Tom,

I have meanwhile moved to using ffmpeg directly. May I humbly suggest to clarify the readme, though? The second sentence in the introduction clearly advertizes the `NVPIPE_H264_FFMPEG` backend as a fallback for when no hardware acceleration is available. If I had known that this is not actually supported that would have saved me valuable time (including the time it took to write the patch in #8 and the two bug reports) and I presume other users will feel the same. 

Cheers,

Rene

**Comment by tfogal** [comment]
Ahh, I had misinterpreted your bug, then; I thought you were using pre-Kepler hardware and trying to get hardware acceleration.  I understand now.

Looking at the code, EOVERFLOW only happens when the output buffer is smaller than the output frame size.  What does `f.size()` evaluate to your case?

**Comment by renewagner** [comment]
Hi Tom,

I'm sorry about the late response. I didn't have access to my test setup until now.

My use case is what I believe is a typical use of NvPipe: A CUDA/OpenGL visualization application running on a GPU server and one or multiple visualization clients receiving rendered frames encoded in H.264 via TCP. The file-based transfer was just for testing purposes.

The test input [test-input.tar.gz](https://github.com/NVIDIA/NvPipe/files/1571975/test-input.tar.gz) consists of a single H.264 encoded frame. `f.size()` evaluates to 103354. I don't think that's part of the problem, though. What happens is that with the `NVPIPE_H264_NV` encoder the frame size is rounded up to an integer multple of 16, i.e. the height is 1088 pixels rather than the original 1080. Since the height is included in the encoded frame the `NVPIPE_H264_FFMPEG` decoder correctly detects that the decoded 1920x1088 frame won't fit into the smaller `clientBuffer`. The `NVPIPE_H264_NV` decoder somehow isn't affected by this.

If I change the `height` parameter in the above code to 1088 the program runs without any errors. Unfortunately, the output PPM is completely black.

I'm not sure how to proceed with this. Now that my purely FFMPEG-based decoder is working I'm afraid I cannot presently devote more time to this, i.e. I need to get back to my actual science problem... Let me know if there are any quick tests I can perform. Otherwise please feel free to close this.

Sorry,

Rene

**Comment by tbiedert** [comment]
Hi René,

this is very likely caused by the internal padding to multiples of 16 which I implemented in commit 60ed7d8570d0d0ce9a055efe9287c4468d1dd5f1. The padding is automatically performed within the color conversion kernels which are part of the NV encoder/decoder. Unfortunately, the FFmpeg-based fallback was a bit neglected along the way, so interop is likely to fail. I'm not quite sure why the PPM turns out completely black, though. I'd just expect a border of 8 pixels at the bottom/top of the output image, as during padded encoding pixels are simply replicated (think of GL_CLAMP_TO_EDGE).

To verify this, you could test your code with one of the previous commits. However, you will probably still need to apply (some of) your patches from the other issue. Alternatively, you could simply render at 1920x1088 and feed that resolution into NvPipe at both NV-encode and FFmpeg-decode side.

Sorry for the inconvenience,
Tim

**Comment by tbiedert** [comment]
I've just pushed a redesigned NvPipe. However, the FFmpeg backend has been deprecated I'm afraid. The stream produced by NvPipe is standard-compliant H.264 or HEVC, so I suggest using FFmpeg for decoding directly in your case.
